src/collective/solr/solr.py

```diff
@@ -212,7 +214,31 @@ def deleteByQuery(self, query):
         xstr = '<delete><query>%s</query></delete>' % self.escapeVal(query)
         return self.doUpdateXML(xstr)
 
+    def get_schema(self):
+        """Memoized access to Solr Schema.
+
+        We need to know which one is the uniqueKey in the add() method below,
+        but don't want to hit Solr with a request for the schema on every
+        single add().
+        """
+        if not hasattr(self, '_schema'):
+            self._schema = self.getSchema()
+        return self._schema
+
     def add(self, boost_values=None, **fields):
+        solr_config = queryUtility(ISolrConnectionConfig)
+        atomic_updates_enabled = getattr(solr_config, 'atomic_updates', True)
+
+        schema = self.get_schema()
+        uniqueKey = schema.get('uniqueKey', None)
+        if uniqueKey is None:
+            raise Exception("Could not get uniqueKey from Solr schema")
```

Comment:
Plain `Exceptions` are a pain to deal with - we should probably introduce a common base class `SolrException` in `collective/solr/exceptions.py` (and make the existing `SolrInactiveException` inherit from that as well).